File: 2020_solutions/cplusplus/main/day9.cpp
#include "day9.hpp"
#include <deque>
#include <algorithm>
#include <numeric>
//
// Created by luker on 2/2/2021.
//

namespace Day9 {
    uint32_t toUint32_t(const std::string string) {
        return static_cast<uint32_t>(std::stoul(string));
    }

    bool findSum(const uint32_t value, const std::deque<uint32_t> &preambleValues) {
        for (auto val1Iter = preambleValues.begin();
             val1Iter != preambleValues.end();
             val1Iter++) {
            for (auto val2Iter = preambleValues.begin();
                 val2Iter != preambleValues.end();
                 val2Iter++) {
                if (val1Iter != val2Iter) {
                    const auto sum = *val1Iter + *val2Iter;
                    if (sum == value) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    uint32_t puzzle1(const size_t preamble, const std::vector<std::string> &inputValues) {
        std::vector<uint32_t> values(inputValues.size());
        std::transform(inputValues.begin(), inputValues.end(), values.begin(), *toUint32_t);
        std::deque<uint32_t> preambleValues;

        for (size_t i = 0; i < preamble; i++) {
            preambleValues.push_back(values[i]);
        }

        for (size_t i = preamble; i < values.size(); i++) {
            uint32_t val = values[i];
            const auto found_sum = findSum(val, preambleValues);
            if (found_sum) {
                preambleValues.pop_front();
                preambleValues.push_back(val);
            } else {
                return val;
            }
        }

        return 0;
    }

    uint32_t puzzle2(const size_t preamble, const std::vector<std::string> &inputValues) {
        std::vector<uint32_t> values(inputValues.size());
        std::transform(inputValues.begin(), inputValues.end(), values.begin(), *toUint32_t);
        const uint32_t invalidNumber = puzzle1(preamble, inputValues);
        const auto invalidNumberIter = std::find(values.begin(), values.end(), invalidNumber);
        const auto invalid_number_index = std::distance(values.begin(), invalidNumberIter);

        for (auto size = 2; size < 20; size++) {
            for (auto start_index = 0; start_index < (invalid_number_index - size); start_index++) {
                    const uint32_t sum = std::accumulate(values.begin() + start_index, values.begin() + start_index + size, 0);

                    if (sum == invalidNumber) {
                        const auto range_values = std::minmax_element(values.begin() + start_index, values.begin() + start_index + size);

                        return *range_values.second + *range_values.first;
                    }

                }

        }

        return 0;
    }
}

File: 2020_solutions/rust/src/day9.rs
use std::collections::VecDeque;
use std::ops::Index;

fn find_sum(value: u64, preamble_values: &VecDeque<u64>) -> Option<u64> {
    for val1 in preamble_values.iter().enumerate() {
        for val2 in preamble_values.iter().enumerate() {
            if (val1.0 != val2.0) {
                let sum = val1.1 + val2.1;
                if (sum == value) {
                    return Option::Some(sum);
                }
            }
        }
    }
    return Option::None;
}

fn puzzle1(preamble: usize, input: Vec<String>) -> u64 {
    let mut values = Vec::<u64>::with_capacity(input.len());
    input.iter()
        .map(|stng|  stng.parse::<u64>().unwrap())
        .for_each(|val| values.push(val));


    let mut preamble_values = VecDeque::<u64>::with_capacity(preamble);

    for val in values.iter().take(preamble) {
        preamble_values.push_back(*val);
    }

    for val in values.iter().skip(preamble) {
        let found_sum = find_sum(*val, &preamble_values);
        match found_sum {
            None => return *val,
            Some(sum) => {
                preamble_values.pop_front();
                preamble_values.push_back(*val);
            }
        }
    }

    return 0;
}

fn puzzle2(preamble: usize, input: Vec<String>) -> u64 {
    let mut values = Vec::<u64>::with_capacity(input.len());
    input.iter()
        .map(|stng|  stng.parse::<u64>().unwrap())
        .for_each(|val| values.push(val));


    let invalid_number = puzzle1(preamble, input);

    let invalid_number_index = values.iter().position(|number| *number == invalid_number).unwrap();

    for size in 2..20 as usize {

        for start_index in 0..(invalid_number_index - size) {

            let sum: u64 = values.iter().skip(start_index)
                .take(size)
                .sum();

            if (sum == invalid_number) {
                let max_range_value = values.iter().skip(start_index)
                    .take(size)
                    .max().unwrap();

                let min_range_value = values.iter().skip(start_index)
                    .take(size)
                    .min().unwrap();

                return *max_range_value + *min_range_value;
            }

        }

    }


    return 0;
}

#[cfg(test)]
mod tests {
    use crate::day9::{puzzle1, puzzle2};
    use crate::utils::read_lines;

    #[test]
    fn test_puzzle_1_1() {
        let strs = vec!["35",
                        "20",
                        "15",
                        "25",
                        "47",
                        "40",
                        "62",
                        "55",
                        "65",
                        "95",
                        "102",
                        "117",
                        "150",
                        "182",
                        "127",
                        "219",
                        "299",
                        "277",
                        "309",
                        "576"];

        let strings = strs.into_iter()
            .map(String::from)
            .collect();

        let result = puzzle1(5, strings);
        assert_eq!(result, 127);
    }

    #[test]
    fn test_puzzle_1_2() {
        let strs = read_lines("data/Day9.txt").unwrap();

        let result = puzzle1(25,strs);
        assert_eq!(result, 556543474);
    }

    #[test]
    fn test_puzzle_2_1() {
        let strs = vec!["35",
                        "20",
                        "15",
                        "25",
                        "47",
                        "40",
                        "62",
                        "55",
                        "65",
                        "95",
                        "102",
                        "117",
                        "150",
                        "182",
                        "127",
                        "219",
                        "299",
                        "277",
                        "309",
                        "576"];

        let strings = strs.into_iter()
            .map(String::from)
            .collect();

        let result = puzzle2(5, strings);
        assert_eq!(result, 62);
    }

    #[test]
    fn test_puzzle_2_2() {
        let strs = read_lines("data/Day9.txt").unwrap();

        let result = puzzle2(25,strs);
        assert_eq!(result, 76096372);
    }
}